common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/WebViewAuthorizationFragment.java

```diff
@@ -140,6 +148,87 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
         }
     }
 
+    @Override
+    public void onResume() {
+        super.onResume();
+        final String methodTag = TAG + ":onResume";
+        final Bundle extras = getExtras();
+        if (isSwitchBrowserResumeFlow(extras)) {
+            Logger.info(methodTag, "Resume switch browser protocol on WebView.");
+            resumeSwitchBrowser(extras);
+        }
+    }
+
+    /**
+     * Get the extras from the activity intent.
+     *
+     * @return Bundle with the extras
+     */
+    @NonNull
+    private Bundle getExtras() {
+        final Activity activity = getActivity();
+        if (activity == null) {
+            return Bundle.EMPTY;
+        }
+        final Intent intent = activity.getIntent();
+        if (intent == null) {
+            return Bundle.EMPTY;
+        }
+        final Bundle extras =  intent.getExtras();
+        return extras == null ? Bundle.EMPTY : extras;
+    }
+
+    /**
+     * Check if the extras contains the switch browser code and action uri.
+     * also checks that the fragment is waiting for the switch browser resume endpoint.
+     * if so, it means we are resuming the switch browser flow.
+     *
+     * @param extras Bundle
+     * @return boolean
+     */
+    private boolean isSwitchBrowserResumeFlow(@NonNull final Bundle extras) {
+        final String methodTag = TAG + ":isSwitchBrowserResumeFlow";
+        final boolean containsActionUri = extras.containsKey(
+                AuthenticationConstants.SWITCH_BROWSER.ACTION_URI
+        );
+        final boolean containsCode = extras.containsKey(
+                AuthenticationConstants.SWITCH_BROWSER.CODE
+        );
+        Logger.verbose(
+                methodTag,
+                "action uri: " + containsActionUri +
+                        ", code: " + containsCode +
+                        ", isWaitingForSwitchBrowserResumeEndpoint: " + isWaitingForSwitchBrowserResumeEndpoint);
+        return containsCode && containsActionUri && isWaitingForSwitchBrowserResumeEndpoint;
+    }
+
```

Comment:
switch browser logic spread across different files: this file, SwitchBrowserUriHelper, SwitchBrowserHandler

I think we can just have one object SwitchBrowserHandler, that you can instantiate in this class handler = SwitchBrowserHandler(activity, ..), and pass it on WebViewClient object.

When it's resume flow, call
if (handler.isSwitchBrowserResumeFlow(extras)) {
     handler.resumeFlow(webview, webviewclient)
}


Above is one idea, maybe other could be another object to resume handling (and first object for switch browser)

This fragment class just hold those object(s) and delegates work to them to continue.

Reply:
synced offline and reworked.